File: lana/src/main/scala/com/financialforce/lana/commands/LogView.scala
/*
 * Copyright (c) 2020 FinancialForce.com, inc. All rights reserved.
 */
package com.financialforce.lana.commands

import com.financialforce.lana.Context
import com.financialforce.lana.display.{OpenFileInPackage, WebView}
import com.financialforce.lana.runtime.vscode.WebviewPanel
import io.scalajs.nodejs.buffer.Buffer
import io.scalajs.nodejs.fs.Fs
import io.scalajs.nodejs.path.Path

import scala.scalajs.js
import scala.scalajs.js.annotation.JSGlobal
import scala.util.Try

@JSGlobal
@js.native
class WebViewLogFileRequest extends js.Object {
  val text: js.UndefOr[String] = js.native
  val typeName: js.UndefOr[String] = js.native
  val path: js.UndefOr[String] = js.native
}

class LogFileException(message: String) extends Exception(message)

trait LogView {
  def createView(wsPath: String,
                 context: Context,
                 logName: String,
                 logPath: String,
                 logContents: String): WebviewPanel = {
    val panel = WebView("logFile", "Log: " + logName, Seq())
    panel.webview.html = getViewContent(context, logName, logPath, logContents)
    // handle callbacks
    panel.webview.onDidReceiveMessage({ a: js.Any =>
      {
        val request: WebViewLogFileRequest = a.asInstanceOf[WebViewLogFileRequest]

        if (request.typeName.nonEmpty) {
          val parts = request.typeName.get.split("-")
          val line = if (parts.length > 1) Try(parts(1).toInt).toOption else None
          OpenFileInPackage.openFileForSymbol(wsPath, context, parts.head, line)
        } else {
          request.path.foreach(context.display.showFile)
        }
      }
    }, js.undefined, js.Array())
    panel
  }

  private def getViewContent(context: Context,
                             logName: String,
                             logPath: String,
                             logContents: String): String = {
    val namespaces: Seq[String] = context.namespaces
    val spaRoot: String = Path.join(context.context.extensionPath, "spa", "log-viewer")
    val index: String = Path.join(spaRoot, "index.html")
    val bundle: String = Path.join(spaRoot, "bundle.js")
    val indexSrc: String = string(Fs.readFileSync(index))
    val bundleSrc: String = string(Fs.readFileSync(bundle))
    val htmlWithLog = insertAtToken(indexSrc, "@@logTxt", logContents)
    val htmlWithLogName = insertAtToken(htmlWithLog, "@@name", logName)
    val htmlWithLogPath = insertAtToken(htmlWithLogName, "@@path", logPath)
    val htmlWithNS = insertAtToken(htmlWithLogPath, "@@ns", namespaces.mkString(","))
    val htmlWithBundle = insertAtToken(htmlWithNS, "@@bundle", bundleSrc)
    htmlWithBundle
  }

  private def string(buf: Buffer): String = {
    buf.toString("utf8", 0, buf.length)
  }

  private def insertAtToken(str: String, token: String, insert: String): String = {
    if (str.indexOf(token) > -1) {
      val splits = str.split(token)
      return splits(0) + insert + splits(1)
    }
    str
  }
}

File: src/commands/LogView.ts
/*
 * Copyright (c) 2020 FinancialForce.com, inc. All rights reserved.
 */

import { WebviewPanel } from "vscode";
import { Context } from "../Context";
import { OpenFileInPackage } from "../display/OpenFileInPackage";
import { WebView } from "../display/WebView";
import * as path from "path";
import { promises as fs } from "fs";

interface WebViewLogFileRequest {
  text: string | undefined;
  typeName: string | undefined;
  path: string | undefined;
}

export class LogFileException extends Error {
  constructor(message: string) {
    super(message);
    this.message = message;
    this.name = "LogFileException";
  }
}

export class LogView {
  static async createView(
    wsPath: string,
    context: Context,
    logName: string
  ): Promise<WebviewPanel> {
    const panel = WebView.apply("logFile", "Log: " + logName, []);
    panel.webview.onDidReceiveMessage(
      (msg: any) => {
        const request = msg as WebViewLogFileRequest;

        if (request.typeName) {
          const parts = request.typeName.split("-");
          let line;
          if (parts.length > 1) {
            line = parseInt(parts[1]);
          }
          OpenFileInPackage.openFileForSymbol(wsPath, context, parts[0], line);
        } else if (request.path) {
          context.display.showFile(request.path);
        }
      },
      undefined,
      []
    );

    return panel;
  }

  static async appendView(
    view: WebviewPanel,
    context: Context,
    logName: string,
    logPath: string,
    logContents: string
  ): Promise<WebviewPanel> {
    view.webview.html = await LogView.getViewContent(
      context,
      logName,
      logPath,
      logContents
    );
    return view;
  }

  private static async getViewContent(
    context: Context,
    logName: string,
    logPath: string,
    logContents: string
  ): Promise<string> {
    const namespaces = context.namespaces;
    const logViewerRoot = path.join(
      context.context.extensionPath,
      "log-viewer"
    );
    const index = path.join(logViewerRoot, "index.html");
    const bundle = path.join(logViewerRoot, "dist", "bundle.js");

    const [indexSrc, bundleSrc] = await Promise.all([
      fs.readFile(index, "utf-8"),
      fs.readFile(bundle, "utf-8"),
    ]);

    const toReplace: { [key: string]: string } = {
      "@@logTxt": logContents,
      "@@name": logName,
      "@@path": logPath,
      "@@ns": namespaces.join(","),
      "@@bundle": bundleSrc,
    };

    return indexSrc.replace(
      /@@logTxt|@@name|@@path|@@ns|@@bundle/gi,
      function (matched) {
        return toReplace[matched];
      }
    );
  }
}
